funnel/templates/formlayout.html.jinja2

```diff
@@ -17,3 +28,91 @@
     <script type="text/javascript" src="{{ ASSET_URL }}"></script>
   {%- endassets -%}
 {% endblock %}
+
+{% block layoutscripts %}
+  {%- if autosave %}
+    <script type="text/javascript">
+      $(function() {
+        var typingTimer;
+        var typingWaitInterval = 1000; // wait till user stops typing for one second to send form data
+        var waitingForResponse = false;
+        var lastSavedData = '';
+
+        $('#revision').val() ? $('#autosave-msg').text('These changes have not been published yet.') : '';
+
+        $('#{{ ref_id }}').on('change', function(e) {
+          autosaveForm();
+        });
+
+        $('#{{ ref_id }}').on('keyup', function(e) {
+          if(e.target.value) {
+            if(typingTimer) clearTimeout(typingTimer);
+            typingTimer = setTimeout(autosaveForm, typingWaitInterval);
+          }
+        });
+
+        function autosaveForm() {
+          if(!waitingForResponse && haveDirtyFields()) {
+            $.ajax({
+              type: 'POST',
+              data:  $("#{{ ref_id }}").serialize() + '&autosave=true',
+              dataType: 'json',
+              timeout: 15000,
+              beforeSend: function() {
+                console.log('sending');
+                $('#autosave-msg').text('Autosaving...');
+                lastSavedData = $("#{{ ref_id }}").find(":input:not(:hidden)").serialize();
+                waitingForResponse = true;
+              },
+              success: function (remoteData) {
+                // Todo: Update window.history.pushState for new form
+                console.log('autosaveForm done', remoteData);
+                $('#autosave-msg').text('Changes saved but not published');
+                if(remoteData.revision) {
+                  $('#revision').val(remoteData.revision);
+                }
+                waitingForResponse = false;
+                autosaveForm();
+              },
+              error: function (response) {
+                console.log('error', response);
+                var errorMsg = '';
+                waitingForResponse = false;
+                if (response.readyState === 4) {
+                  if (response.status === 500) {
+                    errorMsg ='Internal Server Error. Please reload and try again.';
+                  } else if(response.status === 400) {
+                    // There is a version mismatch, notify user to reload the page.
+                    waitingForResponse = true;
+                    errorMsg = 'This page has already been edited. Please reload the page.';
+                  }
+                } else {
+                  errorMsg = 'Unable to connect. Please reload and try again.';
+                }
+                $('#autosave-msg').text(errorMsg);
+                window.toastr.error(errorMsg);
+              },
+            });
+          }
+
+          function haveDirtyFields() {
+            var latestFormData = $('#{{ ref_id }}').find(':input:not(:hidden)').serialize();
```

Comment:
@vidya-ram ^

Reply:
There is event to detect user typing into input fields and event to detect form changes for fields like select, radio. Incase user selects browser's autocomplete, both these events are triggered and we end up sending two ajax events. This was added to avoid this.

We are actually comparing two longs strings since `serialize` returns text string in standard URL-encoded notation.